streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -16,171 +16,256 @@
         this.changelogReader = changelogReader;
         this.processId = processId;
         this.logPrefix = logPrefix;
-        this.streamsMetadataState = streamsMetadataState;
-        this.restoreConsumer = restoreConsumer;
         this.taskCreator = taskCreator;
         this.standbyTaskCreator = standbyTaskCreator;
-        this.active = active;
-        this.standby = standby;
-
+        this.builder = builder;
         final LogContext logContext = new LogContext(logPrefix);
 
-        this.log = logContext.logger(getClass());
+        log = logContext.logger(getClass());
 
         this.adminClient = adminClient;
     }
 
+    void setConsumer(final Consumer<byte[], byte[]> consumer) {
+        this.consumer = consumer;
+    }
+
+    public UUID processId() {
+        return processId;
+    }
+
+    InternalTopologyBuilder builder() {
+        return builder;
+    }
+
     public Admin adminClient() {
         return adminClient;
     }
 
-    void createTasks(final Collection<TopicPartition> assignment) {
-        if (consumer == null) {
-            throw new IllegalStateException(logPrefix + "consumer has not been initialized while adding stream tasks. This should not happen.");
-        }
+    public void handleRebalanceStart() {
+        rebalanceInProgress = true;
+    }
+
+    public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                                 final Map<TaskId, Set<TopicPartition>> standbyTasks) {
+        final Map<TaskId, Set<TopicPartition>> activeTasksToCreate = new TreeMap<>(activeTasks);
 
-        if (!assignment.isEmpty() && !assignedActiveTasks.isEmpty()) {
-            resumeSuspended(assignment);
+        final Map<TaskId, Set<TopicPartition>> standbyTasksToCreate = new TreeMap<>(standbyTasks);
+
+        // first rectify all existing tasks
+        final LinkedHashMap<TaskId, RuntimeException> taskCloseExceptions = new LinkedHashMap<>();
+        final Iterator<Task> iterator = tasks.values().iterator();
+        while (iterator.hasNext()) {
+            final Task task = iterator.next();
+            if (activeTasks.containsKey(task.id()) && task.isActive()) {
+                task.resume();
+                activeTasksToCreate.remove(task.id());
+            } else if (standbyTasks.containsKey(task.id()) && !task.isActive()) {
+                task.resume();
+                standbyTasksToCreate.remove(task.id());
+            } else /* we previously task, and we don't have it anymore, or it has changed active/standby state */ {
+                try {
+                    task.closeClean();
+                } catch (final RuntimeException e) {
+                    log.error(
+                        "Failed to close {} cleanly. Attempting to close remaining tasks before re-throwing.",
+                        task
+                    );
+                    taskCloseExceptions.put(task.id(), e);
+                }
+                iterator.remove();
```

Comment:
If we remove the task from the owned list after capturing the thrown exception, then later on the thread level when we handle exception we would not be able to re-try closing it dirty.

Reply:
Maybe we can do the same as below which is to immediately close it as dirty within the task manager before throwing it to the stream thread since we are closing it anyways?